torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+
+    https://github.com/pytorch/pytorch/blob/074c77601114c66df346c5465667ce8cfaa6d22c/torch/csrc/distributed/c10d/ProcessGroup.hpp#L79
+    https://github.com/pytorch/pytorch/blob/074c77601114c66df346c5465667ce8cfaa6d22c/torch/csrc/distributed/c10d/init.cpp#L937
+    '''
+
+  def __init__(self, prefix_store, rank, size, timeout):
+    super().__init__(rank, size)
+    self.prefix_store = prefix_store  # reserved for future use.
+    self.timeout = timeout
+    self._mesh = []
+
+  def _get_reduce_type(self, reduce_op):
+    if reduce_op == dist.ReduceOp.SUM:
+      return xm.REDUCE_SUM
+    elif reduce_op == dist.ReduceOp.PRODUCT:
+      return xm.REDUCE_MUL
+    elif reduce_op == dist.ReduceOp.BAND:
+      return xm.REDUCE_AND
+    elif reduce_op == dist.ReduceOp.BOR:
+      return xm.REDUCE_OR
+    elif reduce_op == dist.ReduceOp.MIN:
+      return xm.REDUCE_MIN
+    elif reduce_op == dist.ReduceOp.MAX:
+      return xm.REDUCE_MAX
+    elif reduce_op == dist.ReduceOp.BXOR:
+      raise NotImplementedError(f'reduce op {reduce_op}')
+    else:
+      raise ValueError(f'Invalid reduce op {reduce_op}')
+
+  def allreduce(self, tensors, all_reduce_options):
+    reduce_type = self._get_reduce_type(all_reduce_options.reduceOp)
+
+    # TODO(junminh): implement all_reduce_options.timeout.
+    xm.all_reduce(reduce_type, tensors, groups=self._mesh)
+    return WorkXla(tensors)
+
+  def allgather(self, output_tensors_list, input_tensors):
+    for input_tensor, output_tensors in zip(input_tensors, output_tensors_list):
+      result = xm.all_gather(input_tensor, groups=self._mesh)
+      for i, slice in enumerate(torch.split(result, input_tensor.shape[0])):
+        output_tensors[i].set_(slice)
+
+    return WorkXla([t for sublist in output_tensors_list for t in sublist])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L1138
+  def broadcast(self, tensors, opts):
+    root_tensor = tensors[opts.rootTensor]
+    root_rank = opts.rootRank
+    if root_rank != self.rank():
+      with torch.no_grad():
+        root_tensor.zero_()
+    xm.all_reduce(xm.REDUCE_SUM, [root_tensor], groups=self._mesh)
+
+    return WorkXla([root_tensor])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2355
+  def reduce_scatter(self, output_tensors, input_tensors_list, opts):
+    for input_tensors, output_tensor in zip(input_tensors_list, output_tensors):
+      # Ensure all inputs have the same shape.
+      last_shape = input_tensors[0].shape
+      for i, t in enumerate(input_tensors[1:]):
+        if last_shape != t.shape:
+          raise ValueError(f"Input {i+1}'s shape is different from input {i}: "
+                           f"{t.shape} vs {last_shape}")
+        last_shape = t.shape
+      input_tensor = torch.cat(input_tensors)
```

Comment:
Just a heads-up. We have seen users complain that `cat` operations can be slow. If you saw the similar problem, we can try to avoid it by detecting whether input tensors are already on contiguous memory space (in this function), or  coalesce communications (in XLA), or allowing users to pass in a single tensor as input (add new API).

Reply:
ACK